src/js/bucketsort.js

```diff
@@ -0,0 +1,99 @@
+function bucketSort(array, vector_length, getVector, compareEquals) {
+    // array: an array of data
+    // vector_length: the length of vectors coming out of getVector (must be uniform)
+    // getVector: a function that takes an element of array and returns an int vector. defaults to identity
+    // compareEquals: an optional standard sort comparator - if specified it is run on the
+    //               results of the final buckets before returning
+    getVector = getVector || function(d) { return d; };
+
+  var current_sorted_array = array;
+  var current_bucket_ranges = [{lower_index_incl: 0, upper_index_excl: array.length}];
+
+  var new_sorted_array, new_bucket_ranges, bucket_range, sorted_result;
+
+  for (var vector_index=0; vector_index<vector_length; vector_index++) {
+    new_sorted_array = [];
+    new_bucket_ranges = [];
+    // sort each bucket range, and collect sorted array and new bucket ranges
+    for (var j=0; j<current_bucket_ranges.length; j++) {
+      bucket_range = current_bucket_ranges[j];
+      sorted_result = bucketSortHelper(
+          current_sorted_array,
+          vector_length,
+          getVector,
+          bucket_range.lower_index_incl,
+          bucket_range.upper_index_excl,
+          vector_index
+      );
+      if ((vector_index === vector_length-1) && compareEquals) {
+          // if we're on the last bucket, sort with compareAtEnd if specified
+          sorted_result.sorted_array.sort(compareEquals);
+      }
+      new_sorted_array = new_sorted_array.concat(sorted_result.sorted_array);
+      new_bucket_ranges = new_bucket_ranges.concat(sorted_result.bucket_ranges);
+    }
+    current_sorted_array = new_sorted_array;
+    current_bucket_ranges = new_bucket_ranges;
+  }
+  return current_sorted_array;
+};
+
+function compare(vector1, vector2) {
+    // utility function - comparator that describes sort order given by bucketSort
+    if (vector1.length !== vector2.length)
+        throw new Error("oncoprintjs:bucketsort:compare vector1.length != vector2.length");
+
+    var ret = 0;
+    // go left to right, return result of first difference
+    for (var i=0; i<vector1.length; i++) {
+        if (vector1[i] < vector2[i]) {
+            ret = -1;
+            break;
+        } else if (vector1[i] > vector2[i]) {
+            ret = 1;
+            break;
+        }
+    }
+    return ret;
+}
+
+function bucketSortHelper(array, vector_length, getVector, sort_range_lower_index_incl, sort_range_upper_index_excl, vector_index) {
+    // returns { sorted_array: d[], bucket_ranges:{lower_index_incl, upper_index_excl}[]}} }
```

Comment:
more specific comment to say that sorted_array only contains elements from within specified range, and that bucket_ranges are in order